# Python统计基础:如何描述数据


在大数据和人工智能时代，数据科学和机器学习已在许多科学技术领域中变得至关重要。 处理数据的必要方面是能够直观地描述，汇总和表示数据。 Python统计资料库是全面，流行和广泛使用的工具，可帮助您处理数据。

在本教程中，您将学习：

* 可以用来描述和总结数据集的数字量

* 如何在纯Python中计算描述性统计信息

* 如何使用可用的Python库获取描述性统计信息

* 如何可视化数据集


[从此下载示例代码](https://realpython.com/python-statistics/)

## 理解描述性统计

描述性统计信息是关于数据的描述和汇总。 它使用两种主要方法：

* 定量方法以数字方式描述和汇总数据。

* 视觉方法通过图表，曲线图，直方图和其他图形来说明数据。

您可以将描述性统计信息应用于一个或多个数据集或变量。 描述和总结单个变量时，就是在进行单变量分析。 当您搜索一对变量之间的统计关系时，就是在进行双变量分析。 同样，多变量分析一次涉及多个变量。

### 措施类型

在本教程中，您将了解以下描述统计中的度量类型：

* 集中趋势告诉您有关数据中心的信息。有用的度量包括均值，中位数和众数。

* 可变性告诉您有关数据传播的信息。有用的度量包括方差和标准偏差。

* 相关性或联合可变性告诉您有关数据集中一对变量之间的关系。有用的度量包括协方差和相关系数。

您将学习如何使用Python了解和计算这些度量。

### 总体和样本

在统计学中，总体是您所感兴趣的所有元素或项目的集合。它们的数量通常很大，这使得它们不适合收集和分析数据。这就是为什么统计学家通常试图通过选择和检查一个有代表性的子集来得出一些关于总体的结论。总体的这个子集称为样本。理想情况下，样本应在令人满意的程度上保持总体的基本统计特征。这样，您就可以使用样本来收集关于总体的结论。

### 异常点
异常点是指与样本或总体中大多数数据有显著差异的数据点。有许多可能导致离群值的原因，但以下是一些导致您离群的原因：

* 数据自然变化

* 观察系统行为的变化

* 数据收集错误

数据收集错误是异常值的一个特别突出的原因。例如，测量仪器或程序的局限性可能意味着无法获得正确的数据。其他错误可能由错误计算、数据污染、人为错误等引起。异常点并没有精确的数学定义。您必须依靠经验、有关感兴趣的主题的知识和常识来确定数据点是否为离群值以及如何处理它。

## 选择Python统计库

有许多Python统计库可供您使用，但是在本教程中，您将学习一些最流行和广泛使用的统计库。

Python的 statistics 是用于描述性统计信息的内置Python库。 如果您的数据集不是太大，或者您不能依赖导入其他库，则可以使用它。

NumPy是用于数字计算的第三方库，已针对使用一维和多维数组进行了优化。它的主要类型是称为ndarray的数组类型。 该库包含许多用于统计分析的例程。

SciPy是基于NumPy的用于科学计算的第三方库。与NumPy相比，它提供了其他功能，包括用于统计分析的scipy.stats。

Pandas是基于NumPy的用于数值计算的第三方库。它擅长处理带有Series对象的带标签的一维（1D）数据和带有DataFrame对象的二维（2D）数据。

Matplotlib是用于数据可视化的第三方库。 与NumPy，SciPy和Pandas结合使用时效果很好。

请注意，在许多情况下，可以使用Series和DataFrame对象代替NumPy数组。 通常，您可能只是将它们传递给NumPy或SciPy统计函数。 此外，可以通过调用.values或.to_numpy（）从Series或DataFrame中将未标记的数据作为np.ndarray对象。

### Python Statistics库入门

内建的Python统计库拥有数量相对较少的最重要的统计函数。官方文档是查找详细信息的宝贵资源。如果仅限于纯Python，那么Python统计库可能是正确的选择。

开始学习NumPy的一个好地方是官方用户指南，特别是快速入门和基础部分。官方参考可以帮助您刷新您的记忆在具体NumPy概念。在阅读本教程时，您可能想查看统计部分和官方scipy。stats也引用。

如果您想学习panda，那么官方入门页面是一个很好的起点。数据结构介绍可以帮助您了解基本的数据类型、系列和DataFrame。同样，优秀的官方介绍性教程旨在为您提供足够的信息，以便开始在实践中有效地使用Pandas。

matplotlib有一个全面的官方用户指南，您可以使用它来深入了解使用这个库的细节。Matplotlib解剖对于想要开始使用Matplotlib及其相关库的初学者来说是一份非常好的资源。


### 描述性统计计算

首先导入您将需要的所有包

In [1]:
import math
import statistics
import numpy
import scipy.stats
import pandas

这些是进行Python统计信息所需的所有软件包。 通常，您不会使用Python的内置数学包，但在本教程中会很有用。 稍后，您将导入matplotlib.pyplot以进行数据可视化。

让我们创建一些数据来处理。 您将从包含一些任意数字数据的Python列表开始：

In [2]:
x = [8.0, 1, 2.5, 4, 28.0]
x_with_nan = [8.0, 1, 2.5, math.nan, 4, 28.0]

In [3]:
x

[8.0, 1, 2.5, 4, 28.0]

In [4]:
x_with_nan

[8.0, 1, 2.5, nan, 4, 28.0]

现在您有了包含nan的列表x和列表x。它们几乎是相同的，不同的是带有nan的x包含一个nan值。当Python统计例程遇到非数字值(nan)时，理解它们的行为是很重要的。在数据科学中，丢失值是很常见的，您通常会用nan替换它们。

注意:如何获得nan值?在Python中，您可以使用以下任何一种方法

* float('nan')
* math.nan
* numpy.nan

In [5]:
math.isnan(numpy.nan), numpy.isnan(math.nan), math.isnan(x_with_nan[3]), numpy.isnan(x_with_nan[3])


(True, True, True, True)

您可以看到这些功能都是等效的。 但是，请记住，比较两个nan值是否相等会返回False。 换句话说，math.nan == math.nan是False！


现在，创建与x和x_with_nan对应的np.ndarray和pd.Series对象：


In [6]:
y, y_with_nan = numpy.array(x), numpy.array(x_with_nan)
z, z_with_nan = pandas.Series(x), pandas.Series(x_with_nan)
y, y_with_nan,z,z_with_nan

(array([ 8. ,  1. ,  2.5,  4. , 28. ]),
 array([ 8. ,  1. ,  2.5,  nan,  4. , 28. ]),
 0     8.0
 1     1.0
 2     2.5
 3     4.0
 4    28.0
 dtype: float64,
 0     8.0
 1     1.0
 2     2.5
 3     NaN
 4     4.0
 5    28.0
 dtype: float64)

现在有两个NumPy数组(带有nan的是y和y)和两个panda系列(带有nan的是z和z)。所有这些都是一维值序列。

**注意：尽管在本教程中将使用列表，但请记住，在大多数情况下，您可以以相同的方式使用元组。**  
您可以选择为z和z_with_nan中的每个值指定标签。  

## 集中趋势测度

集中趋势的度量显示数据集的集中值或中间值。 对于数据集的中心，有几种定义。 在本教程中，您将学习如何识别和计算这些集中趋势度量：  
* 算数平均值  
* 加权平均值  
* 几何平均值  
* 调谐平均值  
* 中位值
* 众数  

### 算数平均值

样本均值，也称为样本算术平均值或简称为平均值，是数据集中所有项目的算术平均值。
数据集x的平均值在数学上表示为${\frac{\sum_{i=1}^{n}{x_i}}{𝑛}}$，其中𝑖= 1，2，…，𝑛。换句话说，它是所有元素sum的总和除以数据集𝑥中的项目数。  
您可以使用sum（）和len（）使用纯Python计算平均值，而无需导入库：

In [7]:
mean_ = sum(x) / len(x)
mean_


8.7

尽管这很简洁，但是您也可以应用内置的Python统计功能：

In [8]:
mean_ = statistics.mean(x)
mean_

8.7

In [9]:
mean_ = statistics.fmean(x)
mean_

AttributeError: module 'statistics' has no attribute 'fmean'

您已经从内置的Python统计库中调用了mean()和fmean()函数，并得到了与使用纯Python时相同的结果。fmean()是在Python 3.8中引入的，作为比mean()更快的替代方法。它总是返回一个浮点数。但是，如果数据中有nan值，那么statistics.mean()和statistics.fmean()将返回nan作为输出

In [10]:
mean_ = statistics.mean(x_with_nan)
mean_

nan

In [11]:
mean_ = statistics.fmean(x_with_nan)
mean_

AttributeError: module 'statistics' has no attribute 'fmean'

这个结果与sum()的行为一致，因为sum(带有nan的x)也返回nan。如果使用NumPy，则可以通过np.mean()得到均值

In [13]:
mean_ = numpy.mean(y)
mean_

8.7

在上面的示例中，mean()是一个函数，但是您也可以使用相应的方法.mean()

In [14]:
mean_ = y.mean()
mean_

8.7

NumPy中的mean()函数和.mean()方法与statistics.mean()返回相同的结果。如果数据中有nan值，情况也是如此

In [16]:
numpy.mean(y_with_nan), y_with_nan.mean()

(nan, nan)

因此，您通常不需要获得nan值。如果您宁愿忽略nan值，那么您可以使用np.nanmean()


In [18]:
numpy.nanmean(y_with_nan)

8.7

nanmean()忽略了所有nan值。如果要将其应用于没有nan值的数据集，则它将返回与mean()相同的值。pandas.Series对象还有.mean()方法

In [19]:
z_with_nan.mean()

8.7

此行为是可选参数skipna的默认值造成的。您可以通过更改此参数来修改行为。

### 加权平均值

加权平均值也称为加权算术平均值，它是更广义的算术平均值，使我们能够定义每个数据点对结果的相对贡献。 
我们可以为数据集x的每个数据点定义一个权重𝑥i，其中𝑖= 1、2，…，𝑛，𝑛是数据集的项数。然后，将每个数据点与相应的权重相乘，对所有乘积求和，
然后将获得的总和除以权重之和：${\frac{\sum_{i=i}^{n}{w_i}{x_i}}{\sum_{i=i}^{n}{w_i}}}$。算数平均值可以被视为权重均相等的加权平均值。   
**注意：所有权重均为非负值，𝑤ᵢ≥0，且总和等于1，或 ${\sum_{i=1}^{n}{w_i}}$，这很方便（通常是这种情况）**   
当需要包含以给定相对频率出现的项的数据集的平均值时，加权平均值非常方便。例如，假设有一个集合其中20%的元素等于2,50%的元素等于4，剩下的30%的元素等于8。可以计算这样一个集合的均值：

In [20]:
0.2 * 2 + 0.5 * 4 + 0.3 * 8

4.8

这里，频率和权重都被考虑进去了。使用这种方法，不需要知道项目的总数。在纯Python中，可以通过将sum()与range()或zip()结合来实现加权平均值：

In [22]:
x = [8.0, 1, 2.5, 4, 28.0]
w = [0.1, 0.2, 0.3, 0.25, 0.15]
wmean = sum(w[i] * x[i] for i in range(len(x))) / sum(w)
wmean

6.95

In [23]:
wmean = sum(x_ * w_ for (x_, w_) in zip(x, w)) / sum(w)
wmean

6.95

同样，这是一个干净优雅的实现，不需要导入任何库。但是，如果您有大型数据集，那么NumPy可能提供更好的解决方案。可以使用numpy.average()来获得NumPy数组或Pandas系列的加权平均值

In [25]:
y, z, w = numpy.array(x), pandas.Series(x), numpy.array(w)
wmean = numpy.average(y, weights=w)
wmean

6.95

In [27]:
wmean = numpy.average(z, weights=w)
wmean

6.95

结果与纯Python实现的情况相同。 您也可以在普通列表和元组上使用此方法。

另一种解决方案是将按元素乘积w * y与numpy.sum（）或.sum（）结合使用：

In [28]:
w = numpy.array([0.1, 0.2, 0.3, 0.0, 0.2, 0.1])
(w * y_with_nan).sum() / w.sum()

nan

In [29]:
numpy.average(y_with_nan, weights=w), numpy.average(z_with_nan, weights=w)


(nan, nan)

在本例中，average()返回nan，这与numpy.mean()一致。

### 调和平均值

调和平均值是数据集中所有项的倒数平均值的倒数：${\frac{n}{\sum_{i=1}^{n}{\frac{1}{x_i}}}}$，其中𝑖= 1、2，…，𝑛和𝑛是数据集中of的项数。 

调和均值的纯Python的一种实现是：

In [30]:
hmean = len(x) / sum(1 / item for item in x)
hmean

2.7613412228796843

它与相同数据x的算术平均值有很大不同，x的算术平均值是8.7。

以下代码使用statistics.harmonic_mean()计算此调和平均数：

In [31]:
hmean = statistics.harmonic_mean(x)
hmean

2.7613412228796843

上面的示例显示了statistics.harmonic_mean（）的一种应用。如果数据集中具有nan值，则它将返回nan。如果至少有一个0，那么它将返回0。如果数据集中包括负数，那么将返回statistics.StatisticsError。示例如下：

In [32]:
statistics.harmonic_mean(x_with_nan)

nan

In [33]:
 statistics.harmonic_mean([1, 0, 2])

0

In [34]:
statistics.harmonic_mean([1, 2, -2]) 

StatisticsError: harmonic mean does not support negative values

使用此方法时，请记住以下三种情况！

计算调和均值的第三种方法是使用scipy.stats.hmean():

In [35]:
scipy.stats.hmean(y)

2.7613412228796843

In [36]:
scipy.stats.hmean(z)

2.7613412228796843

同样，这是一个非常简单的应用。但是，如果数据集包含nan、0、负数或任何非正数的值，那么结果将得到一个ValueError。

### 集合平均值

几何平均值是数据集𝑥中所有𝑛个元素𝑥ᵢ的乘积的n次方根：${\sqrt[n]{\sum_{i=1}^{n}{x_i}}}$，其中𝑖= 1，2，…，𝑛。 

以下代码用纯Python实现几何平均值：

In [37]:
gmean = 1
for item in x:
    gmean *= item

gmean **= 1 / len(x)
gmean

4.677885674856041

在这种情况下，几何平均值的值与同一数据集x的算术平均值（8.7）和调和平均值（2.76）的值明显不同。

Python 3.8引入了statistics.geometric_mean（），该函数将所有值转换为浮点数并返回其几何平均值：

In [38]:
gmean = statistics.geometric_mean(x)
gmean

AttributeError: module 'statistics' has no attribute 'geometric_mean'

我们得到了与上一个示例相同的结果，但是四舍五入误差极小。如果传递带有nan值的数据，则statistics.geometric_mean()将像大多数类似的函数一样返回nan。



In [ ]:
gmean = statistics.geometric_mean(x_with_nan)
gmean

实际上，这与statistics.mean()、statistics.fmean()和statistics.harmonic_mean()的行为是一致的。如果数据中有一个零或负数，则使用statistics.geometric_mean()将得到statistics.StatisticsError。

还可以通过scipy.stat.gmean()得到几何平均值，以下是示例代码：

In [39]:
scipy.stats.gmean(y)

4.67788567485604

In [40]:
scipy.stats.gmean(z)

4.67788567485604

结果与纯Python实现相同。

如果数据集中有nan值，则gmean()将返回nan。如果至少有一个0，它将返回0.0并给出警告。如果您提供至少一个负数，则会得到nan和警告。

### 中位值

样本中位数是排序数据集的中间元素。数据集可以按升序或降序排序。如果数据集的元素𝑛的数量为奇数，则中位数为中间位置的值：0.5(𝑛+1)。
如果𝑛是偶数，则中位数是中间两个值（即0.5𝑛和0.5𝑛+1处的项）的算术平均值。

例如，如果有数据2、4、1、8和9，则中位数为4，它位于排序后的数据集（1、2、4、8、9）的中间。如果数据点是2、4、1和8，则中位数是3，这是排序序列中两个中间元素（2和4）的平均值。数据集（1、2.5、4、8、28）的中位数为4。如果从数据集中删除异常值28，则中位数成为2.5和4的算术平均值，即3.25。

平均值和中位数的行为之间的主要差异与数据集的异常值或极端值有关。均值受到异常值的严重影响，但中位数受异常值影响比较小或根本不取决于异常值。
考虑数据集（1、2.5、4、8、28），它的平均值是8.7，中值是4。如果异常值28发生变化，假设变成了17，此时均值将减少，而中位值仍然是4；如果异常值28增大，则均值也会增大，但是中位值仍然不变。

这是中位数的许多可能的纯Python实现之一：

In [42]:
n = len(x)
if n % 2:
    median_ = sorted(x)[round(0.5*(n-1))]
else:
    x_ord, index = sorted(x), round(0.5 * n)
    median_ = 0.5 * (x_ord[index-1] + x_ord[index])

median_

4

此实现的两个最重要的步骤如下：

* 排序数据集的元素。
* 在排序的数据集中找到中间元素。

我们也可以使用函数statistics.median()获得中位值：

In [43]:
median_ = statistics.median(x)
median_

4

In [44]:
median_ = statistics.median(x[:-1])
median_

3.25

排序后x的为[1、2.5、4、8.0、28.0]，因此中间的元素为4。x [：-1]排序的结果是[1，2.5，4，8.0]，没有x的最后一项28.0。现在，有两个中间元素，分别为2.5和4。它们的平均值为3.25。

median_low()和median_high() 是另外两个与Python统计资料库中的中位数有关的函数。它们总是从数据集中返回一个元素：

* 如果元素数为奇数，则只有一个中间值，则它们等同于函数mean()。
* 如果元素数为偶数，则有两个中间值。median_low()将返回较小的那个值，median_high()返回较大的那个值。

可通过以下代码比较它们的差异：

In [45]:
statistics.median_low(x[:-1])

2.5

In [46]:
statistics.median_high(x[:-1])

4

x [：-1]排序的结果是[1，2.5，4，8.0]，中间的两个元素是2.5（低）和4（高）。

不同于statistics库的其它函数，当数据集中有nan时，median(), median_low(), 和median_high() 并不返回nan。请看以下代码：

In [47]:
statistics.median(x_with_nan)

6.0

In [48]:
statistics.median_low(x_with_nan)

4

In [49]:
statistics.median_high(x_with_nan)

8.0

**因此，当心这种行为，因为它可能不是您想要的！**

以下代码通过numpy.median()求中位值：

In [51]:
median_ = numpy.median(y)
median_

4.0

In [52]:
median_ = numpy.median(y[:-1])
median_

3.25

我们可以看到numpy.median()的结果与statistics.median()相同。

然而，如果数据集中含有nan时，numpy.median()将引发RuntimeWarning并返回nan。而函数numpy.nanmedian()则会忽略数据集中所有的nan。示例代码如下：

In [53]:
numpy.nanmedian(y_with_nan)

4.0

In [54]:
numpy.nanmedian(y_with_nan[:-1])

3.25

将statistics.median()和numpy.median()应用于数据集x和y获得相同的结果。

Pandas Series对象具有.median()方法，默认情况下会忽略nan值：

In [55]:
z.median()

4.0

In [56]:
z_with_nan.median()

4.0

Pandas对象中的方法.median()的行为与.mean()一致。

以下代码使用可选参数skipna更改此行为。

In [58]:
z_with_nan.median(skipna=False)

nan

### 众数

众数是数据集中出现频率最高的值。如果没有一个这样的值，则该集合是多峰的，因为它具有多众数。例如，在包含点2、3、2、8和12的集合中，2是众数，因为它出现两次，这不同于其它仅出现一次的项。

以下代码是获取众数的Python实现：

In [59]:
u = [2, 3, 2, 8, 12]
mode_ = max((u.count(item), item) for item in set(u))[1]
mode_

2

使用u.count()可以获取u中每个项目的出现次数。 出现次数最多的项目便是众数。

以下代码使用函数statistics.mode()和statistics.multimode()获取众数：

In [60]:
mode_ = statistics.mode(u)
mode_

2

In [61]:
mode_ = statistics.multimode(u)
mode_

AttributeError: module 'statistics' has no attribute 'multimode'

含糊mode()返回单个值，而multimode()返回包含结果的列表。不过，这并不是两个功能之间的唯一区别。如果有多个众数，则mode()会引发StatisticsError，而multimode（）返回具有所有模式的列表。

In [62]:
v = [12, 15, 12, 15, 21, 15, 12]
statistics.mode(v) 

StatisticsError: no unique mode; found 2 equally common values

In [63]:
statistics.multimode(v)

AttributeError: module 'statistics' has no attribute 'multimode'

以上情况应该特别注意，在这两个函数之间进行选择时要特别小心。

statistics.mode()和statistics.multimode()将nan值作为常规值处理，并可以返回nan作为众数。

In [66]:
statistics.mode([2, math.nan, 2])

2

In [67]:
statistics.mode([2, math.nan, 0, math.nan, 5])

nan

在上面的第一个示例中，数字2出现两次并且是模态值。 在第二个示例中，nan是模态值，因为它出现两次，而其他值仅出现一次。

函数scipy.stats.mode()也可以用来求众数，代码如下:

In [69]:
u, v = numpy.array(u), numpy.array(v)
mode_ = scipy.stats.mode(u)
mode_

ModeResult(mode=array([2]), count=array([2]))

In [71]:
mode_ = scipy.stats.mode(v)
mode_

ModeResult(mode=array([12]), count=array([3]))

此函数返回由众数及其出现次数构成的对象。如果数据集中有多个众数，则仅返回最小的那个。

可以使用点标记进一步获取以上NumPy数组中的众数及其其出现的次数，代码如下：

In [74]:
mode_.mode

array([12])

In [75]:
mode_.count

array([3])

此代码使用.mode返回数组v中的最小众数12，并使用.count返回它出现的次数3。scipy.stats.mode()还可以灵活处理nan值。它允许您使用可选参数nan_policy定义所需的行为。此参数可以采用值 'propagate', 'raise' (an error),或'omit'。 

Pandas Series对象具有.mode()方法，该方法可以很好地处理多个众数，并且默认情况下会忽略nan值：

In [78]:
u, v, w = pandas.Series(u), pandas.Series(v), pandas.Series([2, 2, math.nan])
u.mode()

0    2
dtype: int64

In [79]:
v.mode()

0    12
1    15
dtype: int64

In [80]:
w.mode()

0    2.0
dtype: float64

如上所见，.mode()返回一个包含所有众数的新的pd.Series。 如果希望.mode()将nan值考虑在内，则只需传递可选参数dropna = False。

In [81]:
u

0     2
1     3
2     2
3     8
4    12
dtype: int64

In [82]:
v

0    12
1    15
2    12
3    15
4    21
5    15
6    12
dtype: int64